### PR TITLE
AMDGPU: Handle v_add* in eliminateFrameIndex

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102346?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#102346** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102346?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#102345** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102345?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#101694** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/101694?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#101691** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/101691?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#101692](https://github.com/llvm/llvm-project/pull/101692) <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/101692?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by arsenm** [comment]
ping

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp [hunk 1]
```diff
@@ -2452,6 +2452,211 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
...
+      unsigned Src0Idx = NumDefs;
+
+      bool HasClamp = false;
+      MachineOperand *VCCOp = nullptr;
+
+      switch (MI->getOpcode()) {
+      case AMDGPU::V_ADD_U32_e32:
+        break;
+      case AMDGPU::V_ADD_U32_e64:
+        HasClamp = MI->getOperand(3).getImm();
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2470]: Why not to use named operands?

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp [hunk 2]
```diff
@@ -2452,6 +2452,211 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
...
+
+      if (FrameReg && !ST.enableFlatScratch()) {
+        // We should just do an in-place update of the result register. However,
+        // the value there may also be used by the add, in which case we need a
+        // temporary register.
+        //
+        // FIXME: The scavenger is not finding the result register in the
+        // common case where the add does not read the register.
+
+        ScavengedVGPR = RS->scavengeRegisterBackwards(
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2502]: Do I understand it right, if scavenger fails we would have a regression?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2502]: It won't fail, it has the emergency scavenging slot

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2502]: OK, thanks!

**Review by rampitec — APPROVED** [review]
LGTM

**Pull request merged**